File: nodejs-backend/src/services/schedule.service.js
/**
 * Schedule Service
 * Handles schedule-related business logic
 */

const { Schedule, Team } = require('../models');

class ScheduleService {
  /**
   * Get all schedules
   */
  async getAllSchedules() {
    return await Schedule.findAll({
      include: [{
        model: Team,
        as: 'team',
        attributes: ['id', 'name', 'game']
      }],
      order: [['id', 'ASC']]
    });
  }

  /**
   * Get schedule by ID
   */
  async getScheduleById(id) {
    return await Schedule.findByPk(id, {
      include: [{
        model: Team,
        as: 'team',
        attributes: ['id', 'name', 'game']
      }]
    });
  }

  /**
   * Get schedule by day
   */
  async getScheduleByDay(dayOfWeek) {
    return await Schedule.findOne({
      where: { day_of_week: dayOfWeek },
      include: [{
        model: Team,
        as: 'team',
        attributes: ['id', 'name', 'game']
      }]
    });
  }

  /**
   * Get schedules by team ID
   */
  async getSchedulesByTeamId(teamId) {
    return await Schedule.findAll({
      where: { team_id: teamId },
      include: [{
        model: Team,
        as: 'team',
        attributes: ['id', 'name', 'game']
      }],
      order: [['id', 'ASC']]
    });
  }

  /**
   * Create new schedule
   */
  async createSchedule(scheduleData) {
    return await Schedule.create({
      team_id: scheduleData.teamId,
      day_of_week: scheduleData.dayOfWeek,
      start_time: scheduleData.startTime,
      end_time: scheduleData.endTime,
      activity: scheduleData.activity,
      notes: scheduleData.notes
    });
  }

  /**
   * Update schedule
   */
  async updateSchedule(id, scheduleData) {
    const schedule = await Schedule.findByPk(id);
    if (!schedule) {
      throw new Error('Schedule not found with id: ' + id);
    }

    if (scheduleData.dayOfWeek !== undefined) schedule.day_of_week = scheduleData.dayOfWeek;
    if (scheduleData.startTime !== undefined) schedule.start_time = scheduleData.startTime;
    if (scheduleData.endTime !== undefined) schedule.end_time = scheduleData.endTime;
    if (scheduleData.activity !== undefined) schedule.activity = scheduleData.activity;
    if (scheduleData.notes !== undefined) schedule.notes = scheduleData.notes;

    await schedule.save();
    return await this.getScheduleById(schedule.id);
  }

  /**
   * Save or update by day
   */
  async saveOrUpdateByDay(scheduleData) {
    const existing = await Schedule.findOne({
      where: { day_of_week: scheduleData.dayOfWeek }
    });

    if (existing) {
      existing.start_time = scheduleData.startTime;
      existing.end_time = scheduleData.endTime;
      existing.activity = scheduleData.activity;
      existing.notes = scheduleData.notes;
      await existing.save();
      return await this.getScheduleById(existing.id);
    }

    return await this.createSchedule(scheduleData);
  }

  /**
   * Delete schedule
   */
  async deleteSchedule(id) {
    const schedule = await Schedule.findByPk(id);
    if (!schedule) {
      throw new Error('Schedule not found with id: ' + id);
    }

    await schedule.destroy();
    return true;
  }
}

module.exports = new ScheduleService();

File: flutter_app/lib/services/schedule_service.dart
import '../models/schedule.dart';
import 'api_service.dart';

/// Schedule service for managing schedules
class ScheduleService {
  final ApiService _api = ApiService();

  /// Get all schedules
  Future<List<Schedule>> getAllSchedules() async {
    final response = await _api.get('/api/schedules');
    return (response as List).map((json) => Schedule.fromJson(json)).toList();
  }

  /// Get schedule by ID
  Future<Schedule> getScheduleById(int id) async {
    final response = await _api.get('/api/schedules/$id');
    return Schedule.fromJson(response);
  }

  /// Get schedule by day
  Future<Schedule> getScheduleByDay(String day) async {
    final response = await _api.get('/api/schedules/day/$day');
    return Schedule.fromJson(response);
  }

  /// Get schedules by team ID
  Future<List<Schedule>> getSchedulesByTeamId(int teamId) async {
    final response = await _api.get('/api/schedules/team/$teamId');
    return (response as List).map((json) => Schedule.fromJson(json)).toList();
  }

  /// Create schedule (admin only)
  Future<Schedule> createSchedule(Schedule schedule) async {
    final response = await _api.post('/api/schedules', body: schedule.toJson());
    return Schedule.fromJson(response);
  }

  /// Save or update schedule by day (admin only)
  Future<Schedule> saveOrUpdateByDay(Schedule schedule) async {
    final response = await _api.post('/api/schedules/day', body: schedule.toJson());
    return Schedule.fromJson(response);
  }

  /// Update schedule (admin only)
  Future<Schedule> updateSchedule(int id, Schedule schedule) async {
    final response = await _api.put('/api/schedules/$id', body: schedule.toJson());
    return Schedule.fromJson(response);
  }

  /// Delete schedule (admin only)
  Future<void> deleteSchedule(int id) async {
    await _api.delete('/api/schedules/$id');
  }
}
